velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,195 @@
-
-/// This class only implements the basic split version in which the pattern is a
-/// single character
-class SplitCharacter final : public exec::VectorFunction {
+class Split final : public exec::VectorFunction {
  public:
-  explicit SplitCharacter(const char pattern) : pattern_{pattern} {
-    static constexpr std::string_view kRegexChars = ".$|()[{^?*+\\";
-    VELOX_CHECK(
-        kRegexChars.find(pattern) == std::string::npos,
-        "This version of split supports single-length non-regex patterns");
-  }
-
+  Split() {}
   void apply(
       const SelectivityVector& rows,
       std::vector<VectorPtr>& args,
       const TypePtr& /* outputType */,
       exec::EvalCtx& context,
       VectorPtr& result) const override {
-    exec::LocalDecodedVector input(context, *args[0], rows);
-
+    // Get the decoded vectors out of arguments.
+    const bool noLimit = (args.size() == 2);
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* delims = decodedArgs.at(1);
+    DecodedVector* limits = noLimit ? nullptr : decodedArgs.at(2);
     BaseVector::ensureWritable(rows, ARRAY(VARCHAR()), context.pool(), result);
     exec::VectorWriter<Array<Varchar>> resultWriter;
     resultWriter.init(*result->as<ArrayVector>());
-
-    rows.applyToSelected([&](vector_size_t row) {
-      resultWriter.setOffset(row);
-      auto& arrayWriter = resultWriter.current();
-
-      const StringView& current = input->valueAt<StringView>(row);
-      const char* pos = current.begin();
-      const char* end = pos + current.size();
-      const char* delim;
-      do {
-        delim = std::find(pos, end, pattern_);
-        arrayWriter.add_item().setNoCopy(StringView(pos, delim - pos));
-        pos = delim + 1; // Skip past delim.
-      } while (delim != end);
-
-      resultWriter.commit();
-    });
+    int64_t limit = std::numeric_limits<int64_t>::max();
+    // Optimization for the (flat, const, const) case.
+    if (strings->isIdentityMapping() and delims->isConstantMapping() and
+        (noLimit or limits->isConstantMapping())) {
+      const auto* rawStrings = strings->data<StringView>();
+      const auto delim = delims->valueAt<StringView>(0);
+      if (!noLimit) {
+        limit = limits->valueAt<int64_t>(0);
+        if (limit <= 0) {
+          limit = std::numeric_limits<int64_t>::max();
+        }
+      }
+      rows.applyToSelected([&](vector_size_t row) {
+        if (delim.size() == 0) {
+          splitEmptyDelimer(rawStrings[row], limit, row, resultWriter);
+        } else {
+          splitInner(rawStrings[row], delim, limit, row, resultWriter);
+        }
+      });
+    } else {
+      // The rest of the cases are handled through this general path and no
+      // direct access.
+      rows.applyToSelected([&](vector_size_t row) {
+        const auto delim = delims->valueAt<StringView>(row);
+        if (!noLimit) {
+          limit = limits->valueAt<int64_t>(row);
```

Comment:
Consider two cases: 1) delim is constant, limit is not; 2) limit is constant, delim is not. Seems they also go into this else branch, but causes issues in using `->valueAt<>(row)`?

Reply:
yes, it would go to `general` else path but it shouldn't cauases any issues here right ? It just means we don't do optimization for `flat, const, *` or `flat, *, const` cases, i think that should be fine, @PHILO-HE , we can do refactor in following PR if we find that change is needed for perf purpose.